### PR TITLE
ET-3440 Deploy container to aws fargate

### PR DESCRIPTION
**Pull request by carminetoriello**
## What 🕵️ 🔍

- Creates templates for user_api and document_api tasks definitions
- Uses the livedoor docker image for testing the deployment process. The image is attached to the user_api task definition

## References 📝 🔗

- [Jira task](https://xainag.atlassian.net/browse/ET-3440)

## Reviewer checklist

*Reviewer agreement:*

* Reviewers shouldn't review a "Draft Pull Request", but rather wait for it to be marked as "Ready for Review".
* Reviewers do **not** commit or merge the pull request.
* If/when there are changes that need to be applied (applying Terraform or deploying Kubernetes manifests), the **assignee** is the one responsible for doing it.
* Use the labels `not-applied`, `partially-applied` and `fully-applied` so others know the current progress.
  * Remember to change labels accordingly once you start and finish to apply the changes.

### CODE REVIEW

**Comment by jorge-cr-13** [comment]
We need to create one `aws_security_group` group per api. Like this one but using the ports that you created https://github.com/xaynetwork/xayn_infrastructure/blob/master/aws/stage/ecs/main.tf#L63

File: aws/stage/ecs/main.tf [hunk 1]
```diff
@@ -105,3 +113,135 @@ resource "aws_ecs_service" "hello_world" {
...
+
+
+resource "aws_ecs_task_definition" "user_api" {
+  family                   = "user_api_task"
+  network_mode             = "awsvpc"
+  requires_compatibilities = ["FARGATE"]
+  cpu                      = var.container_cpu
+  memory                   = var.container_memory
+  execution_role_arn       = aws_iam_role.ecs_task_execution_role.arn
+  task_role_arn            = aws_iam_role.ecs_task_execution_role.arn
```
Comment (jorge-cr-13) [inline aws/stage/ecs/main.tf:126]: This line is not needed.

File: aws/stage/ecs/main.tf [hunk 2]
```diff
@@ -105,3 +113,135 @@ resource "aws_ecs_service" "hello_world" {
...
+    "cpu": ${var.container_cpu},
+    "memory": ${var.container_memory},
+    "name": "user_api_task",
+    "networkMode": "awsvpc",
+    "logConfiguration": {
+      "logDriver": "awslogs",
+      "options": {
+        "awslogs-group": "${aws_cloudwatch_log_group.ecs_container_insights.id}",
+        "awslogs-region": "eu-central-1",
+        "awslogs-stream-prefix": "ecs-container"
```
Comment (jorge-cr-13) [inline aws/stage/ecs/main.tf:141]: It would be good to give a different prefix to both of the logs.

File: aws/stage/ecs/main.tf [hunk 3]
```diff
@@ -105,3 +113,135 @@ resource "aws_ecs_service" "hello_world" {
...
+]
+DEFINITION
+  runtime_platform {
+    operating_system_family = "LINUX"
+    cpu_architecture        = "X86_64"
+  }
+}
+
+resource "aws_ecs_service" "user_api" {
+  name            = "ecs-service"
```
Comment (jorge-cr-13) [inline aws/stage/ecs/main.tf:166]: Maybe we can call it differently. 
```suggestion
  name            = "ecs-service-user-api"
```

File: aws/stage/ecs/main.tf [hunk 4]
```diff
@@ -105,3 +113,135 @@ resource "aws_ecs_service" "hello_world" {
...
+  cluster         = aws_ecs_cluster.this.id
+  task_definition = aws_ecs_task_definition.user_api.arn
+  desired_count   = 2
+  launch_type     = "FARGATE"
+
+  network_configuration {
+    security_groups = [aws_security_group.ecs_task.id]
```
Comment (jorge-cr-13) [inline aws/stage/ecs/main.tf:173]: We need to use the specific security group for this api.

File: aws/stage/ecs/main.tf [hunk 5]
```diff
@@ -105,3 +113,135 @@ resource "aws_ecs_service" "hello_world" {
...
+    subnets         = data.terraform_remote_state.networking.outputs.subnet_private_ids.*
+  }
+
+  load_balancer {
+    target_group_arn = data.terraform_remote_state.load_balancer.outputs.alb_target_group
+    container_name   = "user_api_task"
```
Comment (jorge-cr-13) [inline aws/stage/ecs/main.tf:179]:
```suggestion
    container_name   = "user_api"
```

File: aws/stage/ecs/main.tf [hunk 6]
```diff
@@ -105,3 +113,135 @@ resource "aws_ecs_service" "hello_world" {
     container_port   = data.aws_ssm_parameter.de_port.value
   }
 }
+
+
+
+resource "aws_ecs_task_definition" "user_api" {
+  family                   = "user_api_task"
```
Comment (jorge-cr-13) [inline aws/stage/ecs/main.tf:120]:
```suggestion
  family                   = "user_api"
```

File: aws/stage/ecs/main.tf [hunk 7]
```diff
@@ -105,3 +113,135 @@ resource "aws_ecs_service" "hello_world" {
...
+
+  load_balancer {
+    target_group_arn = data.terraform_remote_state.load_balancer.outputs.alb_target_group
+    container_name   = "user_api_task"
+    container_port   = data.aws_ssm_parameter.user_api_port.value
+  }
+}
+
+resource "aws_ecs_task_definition" "document_api" {
+  family                   = "document_api_task"
```
Comment (jorge-cr-13) [inline aws/stage/ecs/main.tf:185]:
```suggestion
  family                   = "document_api"
```

File: aws/stage/ecs/main.tf [hunk 8]
```diff
@@ -105,3 +113,135 @@ resource "aws_ecs_service" "hello_world" {
...
+
+resource "aws_ecs_service" "document_api" {
+  name            = "ecs-service"
+  cluster         = aws_ecs_cluster.this.id
+  task_definition = aws_ecs_task_definition.document_api.arn
+  desired_count   = 2
+  launch_type     = "FARGATE"
+
+  network_configuration {
+    security_groups = [aws_security_group.ecs_task.id]
```
Comment (jorge-cr-13) [inline aws/stage/ecs/main.tf:238]: Here we need to use the specific security group

File: aws/stage/ecs/main.tf [hunk 9]
```diff
@@ -105,3 +113,135 @@ resource "aws_ecs_service" "hello_world" {
...
+    subnets         = data.terraform_remote_state.networking.outputs.subnet_private_ids.*
+  }
+
+  load_balancer {
+    target_group_arn = data.terraform_remote_state.load_balancer.outputs.alb_target_group
+    container_name   = "document_api_task"
```
Comment (jorge-cr-13) [inline aws/stage/ecs/main.tf:244]:
```suggestion
    container_name   = "document_api"
```

File: aws/stage/ecs/main.tf [hunk 10]
```diff
@@ -105,3 +113,135 @@ resource "aws_ecs_service" "hello_world" {
...
+      "logDriver": "awslogs",
+      "options": {
+        "awslogs-group": "${aws_cloudwatch_log_group.ecs_container_insights.id}",
+        "awslogs-region": "eu-central-1",
+        "awslogs-stream-prefix": "ecs-container"
+      }
+    },
+    "environment": [
+        {
+          "name": "${data.aws_ssm_parameter.de_port.name}",
```
Comment (jorge-cr-13) [inline aws/stage/ecs/main.tf:211]: I think it is better to use a regular string here.
```suggestion
          "name": "DE_PORT",
```

File: aws/stage/ecs/main.tf [hunk 11]
```diff
@@ -105,3 +113,135 @@ resource "aws_ecs_service" "hello_world" {
...
```
Comment (jorge-cr-13) [inline aws/stage/ecs/main.tf:146]:
```suggestion
          "name": "DE_PORT",
```

**Comment by jorge-cr-13** [comment]
I think the ports here `"aws_ecs_task_definition" "hello_world"` should stay hard coded, anyways we will delete this resources afterwards.

File: aws/stage/ecs/main.tf [hunk 12]
```diff
@@ -98,6 +110,138 @@ resource "aws_ecs_service" "hello_world" {
...
+
+
+resource "aws_ecs_task_definition" "user_api" {
+  family                   = "user_api_task"
+  network_mode             = "awsvpc"
+  requires_compatibilities = ["FARGATE"]
+  cpu                      = var.container_cpu
+  memory                   = var.container_memory
+  execution_role_arn       = aws_iam_role.ecs_task_execution_role.arn
+  task_role_arn            = aws_iam_role.ecs_task_execution_role.arn
```
Comment (jorge-cr-13) [inline aws/stage/ecs/main.tf:126]:
```suggestion
```
This line is not needed.

File: aws/stage/ecs/main.tf [hunk 13]
```diff
@@ -98,6 +110,138 @@ resource "aws_ecs_service" "hello_world" {
...
+}
+
+resource "aws_ecs_task_definition" "document_api" {
+  family                   = "document_api_task"
+  network_mode             = "awsvpc"
+  requires_compatibilities = ["FARGATE"]
+  cpu                      = var.container_cpu
+  memory                   = var.container_memory
+  execution_role_arn       = aws_iam_role.ecs_task_execution_role.arn
+  task_role_arn            = aws_iam_role.ecs_task_execution_role.arn
```
Comment (jorge-cr-13) [inline aws/stage/ecs/main.tf:191]:
```suggestion
```

File: aws/stage/ecs/main.tf (hunk 2 above)
Comment (carminetoriello) [reply aws/stage/ecs/main.tf]: Done!

File: aws/stage/ecs/main.tf (hunk 3 above)
Comment (carminetoriello) [reply aws/stage/ecs/main.tf]: Done!

File: aws/stage/ecs/main.tf (hunk 5 above)
Comment (carminetoriello) [reply aws/stage/ecs/main.tf]: Done!

File: aws/stage/ecs/main.tf (hunk 6 above)
Comment (carminetoriello) [reply aws/stage/ecs/main.tf]: Done!

File: aws/stage/ecs/main.tf (hunk 7 above)
Comment (carminetoriello) [reply aws/stage/ecs/main.tf]: Done!

File: aws/stage/ecs/main.tf (hunk 9 above)
Comment (carminetoriello) [reply aws/stage/ecs/main.tf]: Done!

File: aws/stage/ecs/main.tf (hunk 10 above)
Comment (carminetoriello) [reply aws/stage/ecs/main.tf]: Done!

File: aws/stage/ecs/main.tf (hunk 11 above)
Comment (carminetoriello) [reply aws/stage/ecs/main.tf]: Done!

File: aws/stage/ecs/main.tf (hunk 12 above)
Comment (carminetoriello) [reply aws/stage/ecs/main.tf]: Done!

File: aws/stage/ecs/main.tf (hunk 13 above)
Comment (carminetoriello) [reply aws/stage/ecs/main.tf]: Done!

File: aws/stage/ecs/main.tf (hunk 4 above)
Comment (carminetoriello) [reply aws/stage/ecs/main.tf:210]: Done!

File: aws/stage/ecs/main.tf (hunk 8 above)
Comment (carminetoriello) [reply aws/stage/ecs/main.tf:274]: Done!

File: aws/stage/ecs/main.tf [hunk 14]
```diff
@@ -79,6 +91,44 @@ resource "aws_security_group" "ecs_task" {
   }
 }
 
+resource "aws_security_group" "user_api_task" {
+  name   = "user-api-task-security-group"
+  vpc_id = data.terraform_remote_state.networking.outputs.vpc_id
+
+  ingress {
+    protocol        = "tcp"
+    from_port       = data.aws_ssm_parameter.de_port.value
```
Comment (jorge-cr-13) [inline aws/stage/ecs/main.tf:100]:
```suggestion
    from_port       = data.aws_ssm_parameter.user_api_port.value
```

File: aws/stage/ecs/main.tf [hunk 15]
```diff
@@ -79,6 +91,44 @@ resource "aws_security_group" "ecs_task" {
...
+    to_port         = data.aws_ssm_parameter.de_port.value
```
Comment (jorge-cr-13) [inline aws/stage/ecs/main.tf:101]:
```suggestion
    to_port         = data.aws_ssm_parameter.user_api_port.value
```

File: aws/stage/ecs/main.tf [hunk 16]
```diff
@@ -79,6 +91,44 @@ resource "aws_security_group" "ecs_task" {
...
+    security_groups = [data.terraform_remote_state.load_balancer.outputs.alb_security_group]
+  }
+
+  egress {
+    protocol    = "-1"
+    from_port   = data.aws_ssm_parameter.user_api_port.value
```
Comment (jorge-cr-13) [inline aws/stage/ecs/main.tf:107]:
```suggestion
    from_port   = 0
```

File: aws/stage/ecs/main.tf [hunk 17]
```diff
@@ -79,6 +91,44 @@ resource "aws_security_group" "ecs_task" {
...
+    to_port     = data.aws_ssm_parameter.user_api_port.value
```
Comment (jorge-cr-13) [inline aws/stage/ecs/main.tf:108]:
```suggestion
    to_port     = 0
```

File: aws/stage/ecs/main.tf [hunk 18]
```diff
@@ -79,6 +91,44 @@ resource "aws_security_group" "ecs_task" {
...
+  }
+}
+
+resource "aws_security_group" "document_api_task" {
+  name   = "document-api-task-security-group"
+  vpc_id = data.terraform_remote_state.networking.outputs.vpc_id
+
+  ingress {
+    protocol        = "tcp"
+    from_port       = data.aws_ssm_parameter.de_port.value
```
Comment (jorge-cr-13) [inline aws/stage/ecs/main.tf:119]:
```suggestion
    from_port       = data.aws_ssm_parameter.document_api_port.value
```

File: aws/stage/ecs/main.tf [hunk 19]
```diff
@@ -79,6 +91,44 @@ resource "aws_security_group" "ecs_task" {
...
+    to_port         = data.aws_ssm_parameter.de_port.value
```
Comment (jorge-cr-13) [inline aws/stage/ecs/main.tf:120]:
```suggestion
    to_port         = data.aws_ssm_parameter.document_api_port.value
```

File: aws/stage/ecs/main.tf [hunk 20]
```diff
@@ -79,6 +91,44 @@ resource "aws_security_group" "ecs_task" {
...
+    security_groups = [data.terraform_remote_state.load_balancer.outputs.alb_security_group]
+  }
+
+  egress {
+    protocol    = "-1"
+    from_port   = data.aws_ssm_parameter.document_api_port.value
```
Comment (jorge-cr-13) [inline aws/stage/ecs/main.tf:126]:
```suggestion
    from_port   = 0
```

File: aws/stage/ecs/main.tf [hunk 21]
```diff
@@ -79,6 +91,44 @@ resource "aws_security_group" "ecs_task" {
...
+    to_port     = data.aws_ssm_parameter.document_api_port.value
```
Comment (jorge-cr-13) [inline aws/stage/ecs/main.tf:127]:
```suggestion
    to_port     = 0
```

File: aws/stage/ecs/main.tf [hunk 22]
```diff
@@ -98,6 +148,136 @@ resource "aws_ecs_service" "hello_world" {
...
+
+resource "aws_ecs_service" "user_api" {
+  name            = "ecs-service-user-api"
+  cluster         = aws_ecs_cluster.this.id
+  task_definition = aws_ecs_task_definition.user_api.arn
+  desired_count   = 2
+  launch_type     = "FARGATE"
+
+  network_configuration {
+    security_groups = [aws_security_group.ecs_task.id]
```
Comment (jorge-cr-13) [inline aws/stage/ecs/main.tf:210]:
```suggestion
    security_groups = [aws_security_group.user_api.id]
```

File: aws/stage/ecs/main.tf [hunk 23]
```diff
@@ -98,6 +148,136 @@ resource "aws_ecs_service" "hello_world" {
...
+
+resource "aws_ecs_service" "document_api" {
+  name            = "ecs-service-document-api"
+  cluster         = aws_ecs_cluster.this.id
+  task_definition = aws_ecs_task_definition.document_api.arn
+  desired_count   = 2
+  launch_type     = "FARGATE"
+
+  network_configuration {
+    security_groups = [aws_security_group.ecs_task.id]
```
Comment (jorge-cr-13) [inline aws/stage/ecs/main.tf:274]:
```suggestion
    security_groups = [aws_security_group.document_api_task.id]
```

File: aws/stage/ecs/main.tf (hunk 22 above)
Comment (carminetoriello) [reply aws/stage/ecs/main.tf]: Done!

File: aws/stage/ecs/main.tf (hunk 23 above)
Comment (carminetoriello) [reply aws/stage/ecs/main.tf]: Done!

**Review by jorge-cr-13 — DISMISSED** [review]

File: aws/stage/service/user/main.tf [hunk 24]
```diff
@@ -0,0 +1,124 @@
...
+  {
+    "image": "xaynetci/xayn_discovery_web_service:livedoor",
+    "cpu": ${var.container_cpu},
+    "memory": ${var.container_memory},
+    "name": "user_api_task",
+    "networkMode": "awsvpc",
+    "logConfiguration": {
+      "logDriver": "awslogs",
+      "options": {
+        "awslogs-group": "${aws_cloudwatch_log_group.ecs_container_insights.id}",
```
Comment (jorge-cr-13) [inline aws/stage/service/user/main.tf:81]:
```suggestion
        "awslogs-group": "${data.terraform_remote_state.ecs.outputs.container_insights}",
```

File: aws/stage/service/document/main.tf [hunk 25]
```diff
@@ -0,0 +1,124 @@
...
+  {
+    "image": "",
+    "cpu": ${var.container_cpu},
+    "memory": ${var.container_memory},
+    "name": "document_api_task",
+    "networkMode": "awsvpc",
+    "logConfiguration": {
+      "logDriver": "awslogs",
+      "options": {
+        "awslogs-group": "${aws_cloudwatch_log_group.ecs_container_insights.id}",
```
Comment (jorge-cr-13) [inline aws/stage/service/document/main.tf:81]:
```suggestion
        "awslogs-group": "${data.terraform_remote_state.ecs.outputs.container_insights}",
```

File: aws/stage/ecs/output.tf [hunk 26]
```diff
@@ -0,0 +1,4 @@
+output "ecs_task_execution_role" {
+  value       = aws_iam_role.ecs_task_execution_role.arn
+  description = "ecs task execution role"
+}
```
Comment (jorge-cr-13) [inline aws/stage/ecs/output.tf:4]:
```suggestion
}

output "container_insights" {
   value       = aws_cloudwatch_log_group.ecs_container_insights.id
}

```

**Review by jorge-cr-13 — DISMISSED** [review]

File: aws/stage/services/document/main.tf [hunk 27]
```diff
@@ -0,0 +1,116 @@
...
+    from_port   = 0
+    to_port     = 0
+    cidr_blocks = ["0.0.0.0/0"]
+  }
+}
+
+resource "aws_ecs_cluster" "this" {
+  name = "ecs-cluster"
+}
+
```
Comment (jorge-cr-13) [inline aws/stage/services/document/main.tf:53]:
```suggestion
```
This should not be here, just on the `ecs`.

File: aws/stage/services/user/main.tf [hunk 28]
```diff
@@ -0,0 +1,116 @@
...
+    from_port   = 0
+    to_port     = 0
+    cidr_blocks = ["0.0.0.0/0"]
+  }
+}
+
+resource "aws_ecs_cluster" "this" {
+  name = "ecs-cluster"
+}
+
```
Comment (jorge-cr-13) [inline aws/stage/services/user/main.tf:53]:
```suggestion
```
Same here.

**Review by jorge-cr-13 — APPROVED** [review]

**Pull request merged**